src/world_benchmark.mojo

```diff
@@ -48,7 +38,7 @@ struct Velocity(ComponentType):
 
 
 fn benchmark_new_entities_10_000() raises:
-    world = World()
+    world = World[Position, Velocity]()
```

Comment:
Can't do an in-depth review before next week, but is this really required with the new implementation?

IMO the requirement to specify all component types at world creation really hurts modularity and decoupling.

Reply:
In any case where you have different model versions that require different components. With your implementation, you will always need to know all possible types in the place where you create the world, potentially adding unused types. Without this requirement, you could just use them where needed.

As an example, see my BEEHAVE implementation [beecs](https://github.com/mlange-42/beecs). It is basically a library, and people who want to extend or alter it just import it and do their business. With the requirement discussed, this would not be possible as soon as new components are required. You would have to change the library code itself, preventing the import and requiring a code copy. Consequently, updating for fixes or improvements to the library would be more tedious then necessary, or you may just miss them entirely.

Alternatively, the library could be written in a way that users have to create and inject the world. But that would burden users with the requirement for the respective knowledge.

ATM I can't come up with a use case that would be made completely impossible, but certain things would definitely be less convenient and ergonomic than possible.

Reply:
This looks like it would solve the problem. Not sure about recompilation in Mojo, but in Go this would be the case too, as static linking is the default.

Providing both options also sounds good. Actually, I think performance of component registration is not an issue anyway, as it happens only once in each run.

What do you mean by inheritance in this context? The philosophy of ECS is "composition over inheritance", so it would generally not be used for anything related to entities or components. They won't contain any model logic anyway, as this is implemented in systems.

Reply:
For inheritance, I was thinking about extended models. Say, you have a forest model and want to add more functionality to it. You may want to keep all the processes but add new processes and components. Then you could inherit from the original forest model and add further processes without duplicating code. (I am thinking in an OOP framework, but of course one could also do that differently)

Reply:
Yeah, you should definitely do this differently with ECS. Actually, the "original forest model" is just a list of systems (i.e. sub-models). To extend or modify, one would amend or change this list.

Taking `beecs` as an example again, [this](https://github.com/mlange-42/beecs/blob/0eee3be6e419c320800016a3fe67c438bbf31ad1/model/default.go#L15) is the original model. So there is not even a class or type that implements the actual model. It is all just generic, modular and plugged together.

IMO you definitely reached a point where you should **use** an ECS in a little project, to know what you are developing for.

Reply:
Agreed. I will do that as soon as I have the minimal functionality of the ECS.